### PR TITLE
 added [Feature request] Install a runner via the command line #3669 

### PR DESCRIPTION
**Pull request by Ahmed-Al-Balochi**
Hi Lutris Team, 
I added the features to list, install, and uninstall a runner mentioned in issue #3669.
I hope you find this useful.
Thank you.

### CODE REVIEW

File: lutris/runners/runner.py [hunk 1]
```diff
@@ -324,26 +335,32 @@ def install_cli(self):
...
+                print(self.name + " is now installed :)")
+            except RunnerInstallationError as ex:
+                ErrorDialog(ex.message)
 
     def uninstall_runner_cli(runner_name):
         """
         uninstall the runner given in application file located in lutris/gui/application.py
         provided using lutris -u <runner>
         """
         Runner.name = runner_name
```
Comment (AlexanderRavenheart) [inline lutris/runners/runner.py:358]:
```suggestion
        try:
            runner_class = import_runner(runner_name)
            runner = runner_class()
        except InvalidRunner:
            logger.error("Failed to import Runner: %s", runner_name)
            return
```
InvalidRunner must be included in the import from lutris.runners.

File: lutris/runners/runner.py [hunk 2]
```diff
@@ -324,26 +335,32 @@ def install_cli(self):
...
-        Runner().uninstall()
-        print("Runner is uninstalled")
-        exit()
+        if Runner().can_uninstall() is True:
+            Runner().uninstall()
```
Comment (AlexanderRavenheart) [inline lutris/runners/runner.py:360]:
```suggestion
            runner.uninstall()
```

**Review by AlexanderRavenheart — CHANGES_REQUESTED** [review]
In addition to the below mentions, I believe the contents of the new methods you added in lutris/runners/runner.py should actually be in lutris/gui/application.py since these should not be inherited by all the runners, they are just command line handlers for calling the Runner specific methods: is_installed, can_uninstall, install, uninstall, etc.

File: lutris/runners/runner.py [hunk 3]
```diff
@@ -287,6 +287,64 @@ def install_dialog(self):
...
+    def install_wine_cli(self, version):
+        """
+        install the runner provided in prepare_runner_cli()
+        """
+        from lutris.gui.dialogs.download import simple_downloader
+        self.install(version=version, downloader=simple_downloader, callback=None)
+        print("Wine Runner '"+ version + "' is now installed :)")
+
+    def wine_runner_uninstall(version):
+        version = version+"-x86_64"
```
[the remark this answers is not in the document]
Comment (AlexanderRavenheart) [reply lutris/runners/runner.py]: Side note: It appends the arch of the wine version, not a hard-coded suffix. But I checked the list of available versions and all of them are x86_64, so for now appending that suffix is okay, but might need to be reviewed in the future.

File: lutris/runners/runner.py [hunk 4]
```diff
@@ -287,6 +287,64 @@ def install_dialog(self):
...
+        WINE_DIR = os.path.join(settings.RUNNER_DIR, "wine")
+        runner_path = os.path.join(WINE_DIR, version)
+        if os.path.isdir(runner_path):
+            system.remove_folder(runner_path)
+        print("Wine Runner is Removed")
```
[the remark this answers is not in the document]
Comment (AlexanderRavenheart) [reply lutris/runners/runner.py]: Side note: I didn't notice you were using the remove_folder() from lutris/util/system.py. That handles possible errors so nothing else to do here.

**Comment by Ahmed-Al-Balochi** [comment]
Currently working on migrating the methods from runner.py to application.py.

**Review by AlexanderRavenheart — APPROVED** [review]

**Comment by Ahmed-Al-Balochi** [comment]
This feature should be done. But please take a look at the ```install_cli ``` and ```install_wine_cli``` because I made changes to those methods.
Thank you.

File: lutris/gui/application.py [hunk 5]
```diff
@@ -697,22 +696,89 @@ def print_wine_runners(self):
...
+        else:
+            print(f"Specified version of Wine is not installed: {version}. Please check if the Wine Runner and specified version are installed (--list-wine-runners can be used for that), and that the version specified is in the correct format.")
+
+    def install_cli(self, runner_name):
+        """
+        install the runner provided in prepare_runner_cli()
+        """
+
+        runner = import_runner(runner_name)
+        runner_path = os.path.join(settings.RUNNER_DIR, runner_name)
```
Comment (AlexanderRavenheart) [inline lutris/gui/application.py:751]: This should probably be wrapped in try-except, like you did in uninstall_runner_cli bellow.

**Review by AlexanderRavenheart — APPROVED** [review]

**Comment by Ahmed-Al-Balochi** [comment]
@AlexanderRavenheart,
I have resolved the required changes you asked for plus minor adjustments

File: lutris/gui/application.py [hunk 6]
```diff
@@ -627,6 +682,104 @@ def print_steam_folders(self, command_line):
...
+        if os.path.isdir(runner_path):
+            print(f"Wine version '{version}' is already installed.")
+        else:
+            from lutris.gui.dialogs import ErrorDialog
+            from lutris.gui.dialogs.download import simple_downloader
+            try:
+                runner = import_runner("wine")
+                runner().install(downloader=simple_downloader, version=version)
+                print(f"Wine version '{version}' has been installed.")
+            except RunnerInstallationError as ex:
```
Comment (AlexanderRavenheart) [inline lutris/gui/application.py:732]:
```suggestion
            except (InvalidRunner, RunnerInstallationError) as ex:
```
To cover the possible exception raised by import_runner().

File: lutris/gui/application.py [hunk 7]
```diff
@@ -627,6 +682,104 @@ def print_steam_folders(self, command_line):
...
+        if os.path.isdir(runner_path):
+            print(f"'{runner_name}' is already installed.")
+        else:
+            from lutris.gui.dialogs import ErrorDialog
+            from lutris.gui.dialogs.download import simple_downloader
+            try:
+                runner = import_runner(runner_name)
+                runner().install(version=None, downloader=simple_downloader, callback=None)
+                print(f"'{runner_name}' has been installed")
+            except RunnerInstallationError as ex:
```
Comment (AlexanderRavenheart) [inline lutris/gui/application.py:760]:
```suggestion
            except (InvalidRunner, RunnerInstallationError) as ex:
```

**Review by AlexanderRavenheart — COMMENTED** [review]
Change these 2 and it's all good. 👍🏻

File: lutris/gui/application.py (hunk 7 above)
Comment (Ahmed-Al-Balochi) [reply lutris/gui/application.py]: Done 👍🏻

**Comment by Ahmed-Al-Balochi** [comment]
Thank you @AlexanderRavenheart for guiding me throughout this process, I have learned a lot. It means a great deal to me that my first ever code contribution was to Lutris which is a daily application for me. It feels great contributing back to this awesome software and its community.
Thanks again, and I look forward to making more contributions :)

**Review by AlexanderRavenheart — APPROVED** [review]

**Comment by Ahmed-Al-Balochi** [comment]
Hi @AlexanderRavenheart @strycore,
I think this pull request is ready for prime time, can't wait to see it hopefully in the next Lutris release.
Thank you.
Ahmed

**Comment by strycore** [comment]
Looks good! Merging

**Pull request merged**